### PR TITLE
Sawfly AI rework

### LINKED ISSUE #12015: Sawflies aren't nearly as aggro as they should be
### Labels

[BUG][major]

### Description

Sawflies, on multiplayer, have some pretty big issues with target aquisition, and will rarely attack anyone unless they're standing right next to them or getting assaulted. (at which point they're absolutely brutal in combat, maybe a bit too much so.) Some of this is just critter AI, but I feel like it's a significant enough issue to warrant a bug report.

### Steps to reproduce

1.  Spawn sawflies on multiplayer
2. Watch as they don't do much

### Expected Behavior

Similar to on singleplayer, they should approach players and attack them.

### Additional Information & Screenshots
More of an organizational bug than anything. As per last round, bartemeus released something like 28 sawflies, to no casualties. This is a pretty extreme example of their AI being passive, but they aren't behaving nearly up to the standards I'd like- twenty eight hostile critters shouldn't be something for bored secoffs to deal with if they feel like it.

image

### PR DESCRIPTION
**Pull request by amylizzle**
## About the PR 
Reworks Sawfly's AI to use the generic critter AI stuff instead, for maintainability and performance.

## Why's this needed? 
Sawflies were struggling a bit. Should hopefully fix #12015

### CODE REVIEW

**Review by NightmareChamillian — APPROVED** [review]
Everything seems to be in order, though as was mentioned on discord, their retaliatory double attack is intended to be in the attackby. Thanks for doing this.

**Review by NightmareChamillian — APPROVED** [review]
Everything looks good!

File: code/mob/living/critter/sawfly.dm [hunk 1]
```diff
@@ -194,41 +184,68 @@ This file is the critter itself, and all the custom procs it needs in order to f
...
-		if(!isalive(src)) src.set_density(FALSE) //something in the mob life resets density so it has to be below parent
+		if(prob(8) && istype(src.loc, /turf)) communalbeep() //beep only when not in a grenade
+
+
+	seek_target(range) //ai mob critter targetting behaviour - returns a list of acceptable targets
+		if(src.lastattacker && src.retaliate && GET_DIST(src, src.lastattacker) <= range)
+			return list(src.lastattacker)
+		var/targetcount = 0
+		. = list()
+		for (var/mob/living/C in viewers(range, src))
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/sawfly.dm:216]: C?

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
nitpicks

File: code/mob/living/critter/sawfly.dm (hunk 1 above)
Comment (amylizzle) [reply code/mob/living/critter/sawfly.dm:216]: copy pasted code from the original targetting stuff - pretty sure C used to mean critter some copy/pastes ago. Let's say... C for counter-agent :)

File: code/mob/living/critter/sawfly.dm [hunk 2]
```diff
@@ -194,41 +184,68 @@ This file is the critter itself, and all the custom procs it needs in order to f
...
+			if(istype(C, /mob/living/critter/robotic/sawfly))
+				continue
+			if (isintangible(C))
+				continue
+			if(C.mind?.special_role && issawflybuddy(C))
+				if(!(C.weakref in src.friends))
+					boutput(C, "<span class='alert'>[src]'s IFF system silently flags you as an ally! </span>")
+					src.friends += get_weakref(C)
+				continue
+			if(C.job in list( "Head of Security", "Security Officer", "Nanotrasen Security Consultant")) //hopefully this is cheaper than the OR chain I had before
```
Comment (frawhst) [inline code/mob/living/critter/sawfly.dm:228]: Should make this a static list on the type, bit faster that way.
```suggestion
			if(C.job in list("Head of Security", "Security Officer", "Nanotrasen Security Consultant")) //hopefully this is cheaper than the OR chain I had before
```
(also space)

**Review by frawhst — APPROVED** [review]
other than the one thing lgtm

File: code/mob/living/critter/sawfly.dm (hunk 2 above)
Comment (amylizzle) [reply code/mob/living/critter/sawfly.dm]: done :)

**Review by ZeWaka — APPROVED** [review]

**Pull request merged**